lib/engine-jit/src/artifact.rs

```diff
@@ -325,15 +326,34 @@ impl Artifact for JITArtifact {
         &self.func_data_registry
     }
     fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
-        // let mut s = flexbuffers::FlexbufferSerializer::new();
-        // self.serializable.serialize(&mut s).map_err(|e| SerializeError::Generic(format!("{:?}", e)));
-        // Ok(s.take_buffer())
-        let bytes = bincode::serialize(&self.serializable)
-            .map_err(|e| SerializeError::Generic(format!("{:?}", e)))?;
-
         // Prepend the header.
         let mut serialized = Self::MAGIC_HEADER.to_vec();
-        serialized.extend(bytes);
+
+        serialized.resize(32, 0);
+        let mut writable_leb = &mut serialized[16..];
+        let serialized_data = self.serializable.serialize()?;
+        let length = serialized_data.len();
+        leb128::write::unsigned(&mut writable_leb, length as u64).expect("Should write number");
+
+        let align = std::mem::align_of::<SerializableModule>() as u64;
+
+        let offset = pad_and_extend(&mut serialized, &serialized_data, align);
+        assert_eq!(offset, 32);
+
         Ok(serialized)
     }
 }
+
+/// It pads the data with the desired alignment
+pub fn pad_and_extend(prev_data: &mut Vec<u8>, data: &[u8], align: u64) -> u64 {
+    // We assert that align is a power of 2
+    debug_assert_eq!(align & (align - 1), 0);
+    let align = align as usize;
+    let mut offset = prev_data.len();
+    if offset & (align - 1) != 0 {
+        offset += align - (offset & (align - 1));
+        prev_data.resize(offset, 0);
+    }
+    prev_data.extend(data);
+    offset as u64
+}
```

Comment:
This function in general is a great function for unit tests, it's just way too easy to have subtle bugs with logic like this.

Reply:
Added!

Reply:
Done!